wasm_src/carta_computation/carta_computation.cc

```diff
@@ -222,4 +235,83 @@ void generateVertexData(void* dst, size_t dstCapacity, float* srcVertices, int n
     }
 }
 
+float clamp(float d, float min, float max) {
+    if (d!=d)
+    {
+        return min;
+    }
+    const double t = d < min ? min : d;
+    return t > max ? max : t;
+}
+
+float scaleValue(float x, int scaling, float alpha, float gamma) {
+    switch (scaling)
+    {
+    case SQUARE:
+        return x * x;
+    case SQRT:
+        return sqrt(x);
+    case LOG:
+        return clamp(log(alpha * x + 1.0) / log(alpha), 0.0, 1.0);
+    case POWER:
+        return (pow(alpha, x) - 1.0) / alpha;
+    case GAMMA:
+        return pow(x, gamma);
+    default:
+        return x;
+    }
+}
+
+void calculateCatalogSizeDiameter(float* data, size_t N, float min, float max, int sizeMin, int sizeMax, int scaling, int devicePixelRatio, float alpha, float gamma) {
+    float columnMin = scaleValue(min, scaling, alpha, gamma);
+    float columnMax = scaleValue(max, scaling, alpha, gamma);
+    float range = columnMax - columnMin;
+    for (size_t i = 0; i < N; i++)
+    {
+        float v = clamp(*(data + i), min, max);
+        float value = scaleValue(v, scaling, alpha, gamma);
+        *(data + i) = ((value - columnMin) / range * (sizeMax - sizeMin) + sizeMin) * devicePixelRatio;
+    }
+}
+
+void calculateCatalogSizeArea(float* data, size_t N, float min, float max, int sizeMin, int sizeMax, int scaling, int devicePixelRatio, float alpha, float gamma) {
+    float columnMin = scaleValue(min, scaling, alpha, gamma);
+    float columnMax = scaleValue(max, scaling, alpha, gamma);
+    float range = columnMax - columnMin;
+    for (size_t i = 0; i < N; i++)
+    {
+        float v = clamp(*(data + i), min, max);
+        float value = scaleValue(v, scaling, alpha, gamma);
+        *(data + i) = (sqrt((value - columnMin) / range) * (sizeMax - sizeMin) + sizeMin) * devicePixelRatio;
+    }
+}
+
+void calculateCatalogColorMap(float* data, size_t dataSize, bool invert, float min, float max, int scaling, float alpha, float gamma) {
```

Comment:
I'd rather we used `dataMin` and `dataMax`, to avoid confusion with `std::min/max`